### PR TITLE
Example change for SPA

### PR DESCRIPTION
**Pull request by benaadams**
Mock up example for #2

### CODE REVIEW

**Comment by benaadams** [comment]
Example how you might use it for Jquery SPAs: https://gist.github.com/benaadams/f7d1e9136d8c60939467

**Comment by MaxShehovtsov** [comment]
@benaadams Hi Ben, I've created a new branch https://github.com/Microsoft/ApplicationInsights-JS/tree/spaTimings where I ported your changes just to make is compile and pass all tests (your's became outdated). If you're still interested in bringing it in please merge it to your branch so that the pull request gets updated. At this point the feedback I have is to refactor it to reduce code duplication, after it's done I'll be happy to merge it.

Thanks!

**Comment by benaadams** [comment]
@MaxShehovtsov thank you for the sync update

Ignoring whitespace I think its now merged ok: https://github.com/Microsoft/ApplicationInsights-JS/pull/3/files?w=1

File: JavaScript/JavaScriptSDK/Telemetry/Timings.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+        constructor(durationMs: number, networkMs: number, sendMs: number, receiveMs: number, domProcessingMs: number) {
+            this.duration = 0;
+            var total = Timings.getDuration(durationMs);
+            var network = Timings.getDuration(networkMs);
+            var request = Timings.getDuration(sendMs);
+            var response = Timings.getDuration(receiveMs);
+            var dom = Timings.getDuration(domProcessingMs);
+            
+            if (total < Math.floor(network) + Math.floor(request) + Math.floor(response) + Math.floor(dom)) {
```
Comment (MaxShehovtsov) [inline JavaScript/JavaScriptSDK/Telemetry/Timings.ts]: Looks like Timings acts as a container which you fill out and pass to trackPageView. So it shouldn't do any checks here as they are done in PageViewPerformance.

File: JavaScript/JavaScriptSDK/Telemetry/PageViewPerformance.ts [hunk 2]
```diff
@@ -49,43 +49,57 @@ module Microsoft.ApplicationInsights.Telemetry {
...
+                    this.perfTotal = timings.perfTotal;
+                    this.networkConnect = timings.networkConnect;
+                    this.sentRequest = timings.sentRequest;
+                    this.receivedResponse = timings.receivedResponse;
+                    this.domProcessing = timings.domProcessing;
+                }
+            } else {
+
+                var timing = PageViewPerformance.getPerformanceTiming();
+                if (timing) {
```
Comment (MaxShehovtsov) [inline JavaScript/JavaScriptSDK/Telemetry/PageViewPerformance.ts]: This is common for both cases (with and without custom timings) - getting the values from timings "provider" and checking for total (== 0 and >= than a sum of it's parts). Could you pls refactor it?

**Comment by MaxShehovtsov** [comment]
Also looks like unit tests need to be extended to account for timings case.

**Comment by benaadams** [comment]
Should now be reusing same logic, and just acting as overriding container

**Comment by MaxShehovtsov** [comment]
Cool! looking forward to seeing the unit tests! :+1:

**Comment by benaadams** [comment]
Might take a little while, need to learn a few bits around the testing methodology first

**Comment by MaxShehovtsov** [comment]
For this we need just unit tests (we also have E2E tests, perf tests which are not needed for this pull requests). It's under JavaScriptSDK.Tests/CheckinTests. We run it with Chutzpah (from the visual studio), another way is to open the test harness html made up for selenium tests - JavaScriptSDK.Tests/Selenium/Tests.html (in the same folder - checkinTests.ts which compiles all checkin tests together).
Also we had to tweak chutzpah and the tests are compiled "manually" i.e. with powershell - see inside JavaScriptSDK.Tests.csproj, at the end of the proj definition - this is in case you need to dig into it (hopefully not).

Hope this helps

**Comment by benaadams** [comment]
Not realistically going to get round to this.

**Pull request closed without merging**